### DISCUSSION TITLE
Offline Installation?

### DISCUSSION THREAD

**Discussion by teddybouch** [jab/bidict#173] [Q&A] [answered]
Good afternoon,

I received a piece of code that uses bidict, but the environment that I am trying to use it in does not have a working network connection and also doesn't have pip installed for python3. I'm trying to install the package manually, but it looks like the setup.py script is also dependent on an active network connection able to connect to python.org. Is there a way to do an offline installation of this project that I'm missing? Thank you for any help that you can provide!

Andrew Bouchard

**Comment by jab** [comment, marked as answer]
If you already have a bidict distribution on your filesystem, an active network connection is not required to be able to install bidict into a Python environment. If you want to use bidict's setup.py to install bidict, you do need to install bidict's [`setup_requires` dependency](https://github.com/jab/bidict/blob/5cb2a0640324f6d3b0c387f039db94b3a7aafca1/setup.py#L84), [setuptools_scm](https://github.com/pypa/setuptools_scm), first.

You can also skip using bidict's setup.py altogether, and "install" bidict just by copying the "bidict" subdirectory into your PYTHONPATH, as follows (note that PYTHONPATH includes the current working directory):

```
$ curl python.org  # prove we're offline
curl: (6) Could not resolve host: python.org

$ python -c 'import bidict'  # prove we don't have bidict installed yet
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'bidict'

$ cp -r ~/src/bidict/bidict .  # copy the "bidict" subdir into the PYTHONPATH

$ python -c 'import bidict'  # bidict has now been "installed" :)
```

Of course, having to manually manage the packages in your PYTHONPATH is error-prone, so if there's a lot more of this you'd have to do, it might be worth trying to get a package manager like pip working instead.

I think I've hit the limit of what I can help with here, but I hope this was helpful. If you have any further questions about this, I recommend trying to find Python packaging experts who answer questions in communities like http://discuss.python.org and https://pythondiscord.com.
